src/main/java/com/squareup/javapoet/CodeBlock.java

```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
         if (format.charAt(p) != '$') {
           nextP = format.indexOf('$', p + 1);
           if (nextP == -1) nextP = format.length();
+          formatParts.add(format.substring(p, nextP));
         } else {
           checkState(p + 1 < format.length(), "dangling $ in format string %s", format);
-          switch (format.charAt(p + 1)) {
+          /* Rules:
+           * Format must be in the form $[Numeric Index]{N, L, S, T}
+           * OR
+           * in the form ${$, >, <, [, ]}
+           * Numeric index is ONE BASED (for consistency with java formatting)
+           * Numeric indexing must refer to a format argument within the args array
+           * Only progress iteration over the argument iterator if we do not hit an indexed entry
+           */
+          int countOfIndexCharacters = 0;
+          while (Character.isDigit(format.charAt(p + countOfIndexCharacters + 1))) {
```

Comment:
We probably want a much simpler, much dumber set of index characters. `c >= '0' && c <= '9'`. Character.isDigit supports lots of "digit" like characters!